File: Single Number/Single_Number.py
# 第一种思路，使用数学
# 60ms 28.31%
class Solution:
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        return 2 * sum(set(nums)) - sum(nums)

# 第二种思路，异或，线性的时间复杂度
# 异或是满足交互律的
# 40ms 99.09%
class Solution:
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        temp = 0
        for num in nums:
            temp ^= num
        return temp

File: Single Number/Single_Number.cpp
// Runtime: 12 ms, faster than 98.02% of C++ online submissions for Single Number.
// Memory Usage: 9.8 MB, less than 27.98% of C++ online submissions for Single Number.

class Solution 
{
public:
    int singleNumber(vector<int>& nums) 
    {
        int res = 0;
        for (auto number : nums)
            res ^= number;
        return res;
    }
};